Sources/PackageGraph/Resolution/PlatformVersionProvider.swift

```diff
@@ -0,0 +1,109 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2014-2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import struct PackageModel.MinimumDeploymentTarget
+import struct PackageModel.Platform
+import struct PackageModel.PlatformVersion
+import struct PackageModel.SupportedPlatform
+
+/// Merging two sets of supported platforms, preferring the max constraint
+func merge(into partial: inout [SupportedPlatform], platforms: [SupportedPlatform]) {
+    for platformSupport in platforms {
+        if let existing = partial.firstIndex(where: { $0.platform == platformSupport.platform }) {
+            if partial[existing].version < platformSupport.version {
+                partial.remove(at: existing)
+                partial.append(platformSupport)
+            }
+        } else {
+            partial.append(platformSupport)
+        }
+    }
+}
+
+public struct PlatformVersionProvider: Hashable {
+    public enum Implementation: Hashable {
+        case mergingFromTargets([ResolvedTarget])
+        case customXCTestMinimumDeploymentTargets([PackageModel.Platform: PlatformVersion]?)
+        case empty
+    }
+
+    private let implementation: Implementation
+
+    public init(implementation: Implementation) {
+        self.implementation = implementation
+    }
+
+    func derivedXCTestPlatformProvider(_ declared: PackageModel.Platform) -> PlatformVersion? {
+        switch implementation {
+        case .mergingFromTargets(let targets):
+            let platforms = targets.reduce(into: [SupportedPlatform]()) { partial, item in
+                merge(into: &partial, platforms: [item.getSupportedPlatform(for: declared, usingXCTest: item.type == .test)])
+            }
+            return platforms.first!.version
+
+        case .customXCTestMinimumDeploymentTargets(let customXCTestMinimumDeploymentTargets):
```

Comment:
Since we are creating the implementation enum here, I wonder if we should make custom and default two separate cases?

Reply:
Good catch, I also found that `Implementation.empty` case is unused.